### ISSUE TITLE
Evaluation context structured value.

### ISSUE THREAD

**Issue by kinyoklion** [open-feature/dotnet-sdk#25]
Looking at the EvaluationContext I see that you can use any `object` as a value in the context. I am not clear how a provider could make use of non-trivial types, or be able to easily serialize the overall context without some level of interface requirement. (Unless they provided provider specific types, or specified specific types, but that doesn't seem very interoperable.)

In the basic use case, where there are top level basic types (string, number, boolean), I can see how a provider could request something as a type based on the meta-data that they likely have.

In the case of a `structure` use case I am not sure how that could be accomplished. I do see how it could be accomplished if the types were effectively restricted to each of the base types, as well as `ImmutableList` and `ImmutableDictionary` of those those types and more `ImmutableLists` and `ImmutableDictionary`.

Or is there going to be a builder which can mange those constraints?

Imagining a context of this general shape:
```
{
  "name": "example",
  "complex": {
    "field1": "value"
  }
}
```

How could a provider access the value of field1? Aside from using reflection, or using some pre-specified type?

Edit:
The node implementation contains this type, which I think expresses how this data can be made navigable and understood.
```
  export type JSONValue = null | string | number | boolean | Date | { [x: string]: JSONValue } | Array<JSONValue>;
```

Thank you,
Ryan

**Comment by benjiro** [comment]
Thanks for the issue @kinyoklion, sorry I'm pretty tied up with work commitments at the moment so I haven't had a chance to look at this in detail.

The current implementation is crude and needs some work as you've discovered, the main reason I chose not to include some form of serialisation was I didn't want to tie the SDK to a specific library like Newtonsoft, System.Text.Json, etc. I believe this logic should be up to the provider to decide how they want to use it. The interactions in the context do need some work, currently, you can get a specific key using the Get method with the known type but if you are using array accessors you'd need to case the object which isn't ideal.

In your example, I would expect the provider to define a structured object say 

```csharp
public class ComplexObject 
{
    public Name {get;set;}
    public Dictionary<string, string> Complex {get;set;}
}

context.Add("complex-object", new ComplexObject());

var complexObject = context.Get<ComplexObject>("complex-object")
complexObject.complex["field1"]
```

The provider would transform that into the Provider specific SDK, lets take cloudbee/rollout for example which uses a ContextBuilder object. it would look something like

```csharp
// https://docs.cloudbees.com/docs/cloudbees-feature-management-api/latest/api-reference/csharp-api#_interface_icontext
return new ContextBuilder().Build(context?.ToDictionary(x => x.Key, x => x.Value) ?? new Dictionary<string, object>());
```

I'll have a think about this more and try to improve the developer's experience around the interactions with the EvaluationContext. Personally, I want to stay away from reflection due to the performance overhead it entails.

**Comment by kinyoklion** [comment]
I do think  a builder would be a great idea. 

Ultimately I think it is important that the types being removed from the EvaluationContext can be understood enough to translate into whatever form the provider needs. So, even if you use a builder, if the storage or interface contains `object` I think that contents of the context will end up being discarded, or resulting in some error condition.

An example of how this works in the LaunchDarkly dotnet SDK is that the values are constrained to this type. Which contains things that can be represented in JSON. Though something similar could be made without that constraint (primarily allowing more numeric types).
https://github.com/launchdarkly/dotnet-sdk-common/blob/main/src/LaunchDarkly.CommonSdk/LdValue.cs

The things that can be provided though are that arrays and dictionaries are immutable collections of the specified value type. Being as all the types, aside from the array/dictionary, are value types it also grants a level of immutability after a context is constructed. Not all languages can provide this immutability, but when it can be provided it is nice to know that the contents of reference types will not change during the course of the evaluation.

Additionally, being as all types are some combination of known types, it means that serialization can be implemented at the provider level without needing to resort to reflection.

**Comment by toddbaert** [comment]
Here is a very relevant [issue](https://github.com/open-feature/java-sdk/issues/50) and [PR](https://github.com/open-feature/java-sdk/pull/51) in Java. Please take a look at these discussions. I think a similar API would work for .NET.

**Comment by kinyoklion** [comment]
@toddbaert Thanks for pointing that out.

At a glance it does look like a substantial step in the right direction.

One thing I am not sure about is arrays, and possibly I just don't know Java well enough, but is there a constraint that arrays contain the same types as structures, as well as structures themselves?

**Comment by toddbaert** [comment]
> One thing I am not sure about is arrays, and possibly I just don't know Java well enough, but is there a constraint that arrays contain the same types as structures, as well as structures themselves?

I am a bit concerned by this as well, but in Java it's not possible to describe a type union, so it would be tough to restrict `List<T>` where `T` is `Structure | Boolean | Integer... etc `.

**Comment by kinyoklion** [comment]
I suppose the concern with a value wrapper approach would be the intermediate objects? Because then the List is just of the wrapper type instead of T?

https://github.com/launchdarkly/java-sdk-common/blob/main/src/main/java/com/launchdarkly/sdk/LDValue.java

It is a reasonably tricky problem in Java. I think it is a bit more straightforward in C#.

**Comment by toddbaert** [comment]
Personally I think the value-wrapper approach is a _bit_ more trouble than it's worth. The generic structure solution approaches the same robustness with a lot less boilerplate, IMO.

**Comment by kinyoklion** [comment]
It is certainly a lot more code for the library implementer. It does prevent the array type problem from being a runtime issue though. Potentially there is some intermediate when it comes to arrays specifically though. Implementation wise it seems like each value could be annotated like you are doing for the individual values by making them pairs.

Though I am not sure what the interface could be for heterogeneous arrays.

**Comment by benjiro** [comment]
This issue has been resolved under #53
